connector/connect/common/src/main/protobuf/spark/connect/base.proto

```diff
@@ -539,6 +558,138 @@ message AddArtifactsResponse {
   repeated ArtifactSummary artifacts = 1;
 }
 
+message InterruptRequest {
+    // (Required)
+    //
+    // The session_id specifies a spark session for a user id (which is specified
+    // by user_context.user_id). The session_id is set by the client to be able to
+    // collate streaming responses from different queries within the dedicated session.
+    string session_id = 1;
+
+    // (Required) User context
+    UserContext user_context = 2;
+
+    // Provides optional information about the client sending the request. This field
+    // can be used for language or version specific information and is only intended for
+    // logging purposes and will not be interpreted by the server.
+    optional string client_type = 3;
+
+    // (Required) The operation request id to be interrupted.
+    string operation_id = 4;
+}
+
+// TODO - depracated by ReleaseOperationRequest becomes reudndant
+message InterruptResponse {
+  string session_id = 1;
+  string operation_id = 2;
+}
+
+message ReleaseOperationRequest {
+  // (Required)
+  //
+  // The session_id specifies a spark session for a user id (which is specified
+  // by user_context.user_id). The session_id is set by the client to be able to
+  // collate streaming responses from different queries within the dedicated session.
+  string session_id = 1;
+
+  // (Required) User context
+  UserContext user_context = 2;
+
+  // Provides optional information about the client sending the request. This field
+  // can be used for language or version specific information and is only intended for
+  // logging purposes and will not be interpreted by the server.
+  optional string client_type = 3;
+
+  // (Required) The operation request id to be interrupted.
+  string operation_id = 4;
+
+  // When true, all data about the operation can be fully released from the server,
+  // and any executing operation can be interrupted.
+  bool remove = 5;
+
+  // (Optional) If remove is false, this can be used to instruct the server to interrupt the
+  // operation execution.
+  optional bool interrupt = 6;
+
+  // (Optional) If remove is false, this can be used to instruct the server that query
+  // below this offset (exclusive) will no longer be requested by the client.
+  optional int64 free_query_results_below_offset = 7;
```

Comment:
IIUC, the purpose here is  get signal from client -> server about how much the client has consumed of the result stream. 

How would you feel about a separating direct vs. non-direct on Execute followed by a separate bidi streaming API for streaming results?
```
ExecutePlan(mode=DIRECT|ASYNC) -> ExecutePlanResponse(
  one_of=
     streaming_results /* if direct */ 
      | op_id /* if async. No results streamed back here */
)

// If async send this
// Client sends `WatchExecutionReq on the client stream to inform the server about it's progress and it can throw things away. 
// If during streaming results, the server receives a WatchExecutionReq, it rewinds to that offset and re-streams results.
rpc WatchExecution(stream WatchExecutionReq) stream WatchExecutionRes ;
message WatchExecutionReq {
  opId = 1
  offset = 2
}

message WatchExecutionResp{
  oneof response_type {
     same as ExecutePlanResp
  }
}
```